File: golang/lib/kurtosis_context/kurtosis_context.go
package kurtosis_context

import (
	"context"
	"fmt"
	"github.com/kurtosis-tech/kurtosis-engine-api-lib/golang/kurtosis_engine_rpc_api_bindings"
	"github.com/kurtosis-tech/kurtosis-engine-api-lib/golang/kurtosis_engine_rpc_api_consts"
	"github.com/kurtosis-tech/kurtosis-engine-api-lib/golang/lib/api_container_context"
	"github.com/kurtosis-tech/kurtosis-engine-api-lib/golang/lib/enclave_context"
	"github.com/kurtosis-tech/stacktrace"
	"google.golang.org/grpc"
	"google.golang.org/protobuf/types/known/emptypb"
)

const (
	localHostIPAddressStr = "0.0.0.0"
)

type KurtosisContext struct {
	client kurtosis_engine_rpc_api_bindings.EngineServiceClient
}

// NewKurtosisContextFromLocalEngine
// Attempts to create a KurtosisContext connected to a Kurtosis engine running locally
func NewKurtosisContextFromLocalEngine() (*KurtosisContext, error) {
	kurtosisEngineSocketStr := fmt.Sprintf("%v:%v", localHostIPAddressStr, kurtosis_engine_rpc_api_consts.ListenPort)

	// TODO SECURITY: Use HTTPS to ensure we're connecting to the real Kurtosis API servers
	conn, err := grpc.Dial(kurtosisEngineSocketStr, grpc.WithInsecure())
	if err != nil {
		return nil, stacktrace.Propagate(
			err,
			"An error occurred creating a connection to the Kurtosis Engine Server at '%v'",
			kurtosisEngineSocketStr,
		)
	}

	engineServiceClient := kurtosis_engine_rpc_api_bindings.NewEngineServiceClient(conn)

	kurtosisContext := &KurtosisContext{
		client: engineServiceClient,
	}

	return kurtosisContext, nil
}

func (kurtosisCtx *KurtosisContext) CreateEnclave(
	ctx context.Context,
	enclaveId string,
	apiContainerImage string,
	apiContainerLogLevel string,
	isPartitioningEnabled bool,
	shouldPublishPorts bool,
) (*enclave_context.EnclaveContext, error) {

	createEnclaveArgs := &kurtosis_engine_rpc_api_bindings.CreateEnclaveArgs{
		EnclaveId: enclaveId,
		ApiContainerImage: apiContainerImage,
		ApiContainerLogLevel: apiContainerLogLevel,
		IsPartitioningEnabled: isPartitioningEnabled,
		ShouldPublishAllPorts: shouldPublishPorts,
	}

	response, err := kurtosisCtx.client.CreateEnclave(ctx, createEnclaveArgs)
	if err != nil {
		return nil, stacktrace.Propagate(err, "An error occurred creating an enclave with ID '%v'", enclaveId)
	}

	enclaveContext, err := newEnclaveContextFromEnclaveInfo(response.EnclaveInfo)
	if err != nil {
		return nil, stacktrace.Propagate(err, "An error occurred creating an enclave context from a newly-created enclave; this should never happen")
	}

	return enclaveContext, nil
}

func (kurtosisCtx *KurtosisContext) GetEnclaves(ctx context.Context) (map[string]*enclave_context.EnclaveContext, error) {

	response, err := kurtosisCtx.client.GetEnclaves(ctx, &emptypb.Empty{})
	if err != nil {
		return nil,
		stacktrace.Propagate(
			err,
			"An error occurred getting enclaves")
	}

	enclavesMap := newEnclaveContextMapFromEnclaveInfoMap(response.EnclaveInfo)

	return enclavesMap, nil
}

func (kurtosisCtx *KurtosisContext) StopEnclave(ctx context.Context, enclaveId string) error {
	stopEnclaveArgs := &kurtosis_engine_rpc_api_bindings.StopEnclaveArgs{EnclaveId: enclaveId}

	if _, err := kurtosisCtx.client.StopEnclave(ctx, stopEnclaveArgs); err != nil {
		return stacktrace.Propagate(err, "An error occurred stopping enclave with ID '%v'", enclaveId)
	}

	return nil
}

func (kurtosisCtx *KurtosisContext) DestroyEnclave(ctx context.Context, enclaveId string) error {
	destroyEnclaveArgs := &kurtosis_engine_rpc_api_bindings.DestroyEnclaveArgs{
		EnclaveId: enclaveId,
	}

	if _, err := kurtosisCtx.client.DestroyEnclave(ctx, destroyEnclaveArgs); err != nil {
		return stacktrace.Propagate(err, "An error occurred destroying enclave with ID '%v'", enclaveId)
	}

	return nil
}

// ====================================================================================================
// 									   Private helper methods
// ====================================================================================================
func newEnclaveContextMapFromEnclaveInfoMap(
		enclaveInfoMap map[string]*kurtosis_engine_rpc_api_bindings.EnclaveInfo,
		) map[string]*enclave_context.EnclaveContext {

	enclaveContextMap := map[string]*enclave_context.EnclaveContext{}
	for enclaveId, enclaveInfo := range enclaveInfoMap {
		enclaveContext, err := newEnclaveContextFromEnclaveInfo(enclaveInfo)
		if err != nil {
			// TODO This is really nasty - we skip enclaves that we can't create, and the only reason we wouldn't be able to create
			//  enclaves is because of stopped containers. Basically, we should move the API container into the engine-server, and
			//  make it impossible for enclaves to be in an invalid state at all
			continue
		}
		enclaveContextMap[enclaveId] = enclaveContext
	}

	return enclaveContextMap
}

func newEnclaveContextFromEnclaveInfo(
	enclaveInfo *kurtosis_engine_rpc_api_bindings.EnclaveInfo,
) (*enclave_context.EnclaveContext, error) {

	enclaveContainersStatus := enclaveInfo.GetContainersStatus()
	if enclaveContainersStatus != kurtosis_engine_rpc_api_bindings.EnclaveContainersStatus_EnclaveContainersStatus_RUNNING {
		return nil, stacktrace.NewError(
			"Enclave containers status was '%v', but we can't create an enclave context from a non-running enclave",
			enclaveContainersStatus,
		)
	}

	enclaveApiContainerStatus := enclaveInfo.GetApiContainerStatus()
	if enclaveApiContainerStatus != kurtosis_engine_rpc_api_bindings.EnclaveAPIContainerStatus_EnclaveAPIContainerStatus_RUNNING {
		return nil, stacktrace.NewError(
			"Enclave API container status was '%v', but we can't create an enclave context without a running API container",
			enclaveApiContainerStatus,
		)
	}

	apiContainerInfo := enclaveInfo.GetApiContainerInfo()
	if apiContainerInfo == nil {
		return nil, stacktrace.NewError("API container was listed as running, but no API container info exists")
	}
	apiContainerHostMachineInfo := enclaveInfo.GetApiContainerHostMachineInfo()
	if apiContainerHostMachineInfo == nil {
		return nil, stacktrace.NewError("API container was listed as running, but no API container host machine info exists")
	}


	apiContainerContext := api_container_context.NewAPIContainerContext(
		apiContainerInfo.GetContainerId(),
		apiContainerInfo.GetIpInsideEnclave(),
		apiContainerInfo.GetPortInsideEnclave(),
		apiContainerHostMachineInfo.GetIpOnHostMachine(),
		apiContainerHostMachineInfo.GetPortOnHostMachine(),
	)

	enclaveContext := enclave_context.NewEnclaveContext(
		enclaveInfo.GetEnclaveId(),
		enclaveInfo.GetNetworkId(),
		enclaveInfo.GetNetworkCidr(),
		apiContainerContext,
	)
	return enclaveContext, nil
}

File: typescript/src/lib/kurtosis_context/kurtosis_context.ts
import { EngineServiceClient } from "../../kurtosis_engine_rpc_api_bindings/engine_service_grpc_pb";
import { LISTEN_PORT} from "../../kurtosis_engine_rpc_api_consts/kurtosis_engine_rpc_api_consts";
import {EnclaveContext} from "../enclave_context/enclave_context";
import * as grpc from "grpc";
import { Result, err, ok, Err } from "neverthrow";
import {newCreateEnclaveArgs, newDestroyEnclaveArgs, newStopEnclaveArgs} from "../constructor_calls";
import {
    CreateEnclaveArgs,
    CreateEnclaveResponse, DestroyEnclaveArgs, EnclaveAPIContainerHostMachineInfo, EnclaveAPIContainerInfo, EnclaveAPIContainerStatus, EnclaveAPIContainerStatusMap,
    EnclaveContainersStatus,
    EnclaveContainersStatusMap,
    EnclaveInfo, GetEnclavesResponse, StopEnclaveArgs
} from "../../kurtosis_engine_rpc_api_bindings/engine_service_pb";
import {ApiContainerContext} from "../api_container_context/api_container_context";
import * as google_protobuf_empty_pb from "google-protobuf/google/protobuf/empty_pb";
import * as jspb from "google-protobuf";

const LOCAL_HOST_IP_ADDRESS_STR: string = "0.0.0.0";

export class KurtosisContext {
    private readonly client: EngineServiceClient;

    private constructor(client: EngineServiceClient) {
        this.client = client;
    }

    // Attempts to create a KurtosisContext connected to a Kurtosis engine running locally
    public static newKurtosisContextFromLocalEngine(): Result<KurtosisContext, Error>{
        const kurtosisEngineSocketStr: string = `${LOCAL_HOST_IP_ADDRESS_STR}:${LISTEN_PORT}`;

        let engineServiceClient: EngineServiceClient;
        // TODO SECURITY: Use HTTPS to ensure we're connecting to the real Kurtosis API servers
        try {
            engineServiceClient = new EngineServiceClient(kurtosisEngineSocketStr, grpc.credentials.createInsecure());
        } catch(exception) {
            if (exception instanceof Error) {
                return err(exception);
            }
            return err(new Error(
                "An unknown exception value was thrown during creation of the engine client that wasn't an error: " + exception
            ));
        }

        const kurtosisContext: KurtosisContext = new KurtosisContext(engineServiceClient);

        return ok(kurtosisContext);
    }

    public async createEnclave(
        enclaveId: string,
        apiContainerImage: string,
        apiContainerLogLevel: string,
        isPartitioningEnabled: boolean,
        shouldPublishPorts: boolean): Promise<Result<EnclaveContext, Error>> {

        const args: CreateEnclaveArgs = newCreateEnclaveArgs(
            enclaveId,
            apiContainerImage,
            apiContainerLogLevel,
            isPartitioningEnabled,
            shouldPublishPorts)

        const createEnclavePromise: Promise<Result<CreateEnclaveResponse, Error>> = new Promise((resolve, _unusedReject) => {
            this.client.createEnclave(args, (error: grpc.ServiceError | null, response?: CreateEnclaveResponse) => {
                if (error === null) {
                    if (!response) {
                        resolve(err(new Error("No error was encountered but the response was still falsy; this should never happen")));
                    } else {
                        resolve(ok(response!));
                    }
                } else {
                    resolve(err(error));
                }
            })
        });

        const createEnclaveResult: Result<CreateEnclaveResponse, Error> = await createEnclavePromise;
        if (!createEnclaveResult.isOk()) {
            return err(createEnclaveResult.error)
        }

        const response: CreateEnclaveResponse = createEnclaveResult.value;

        const enclaveInfo: EnclaveInfo | undefined = response.getEnclaveInfo();
        if (enclaveInfo === undefined) {
            return err(new Error("An error occurred creating enclave with ID " + enclaveId + " enclaveInfo is undefined; this is a bug on this library" ))
        }
        const newEnclaveContextResult: Result<EnclaveContext, Error> = this.newEnclaveContextFromEnclaveInfo(enclaveInfo);
        if (newEnclaveContextResult.isErr()) {
            return err(new Error(`An error occurred creating an enclave context from a newly-created enclave; this should never happen`))
        }

        return ok(newEnclaveContextResult.value);
    }

    public async getEnclaves(): Promise<Result<Map<string, EnclaveContext>, Error>>{

        const emptyArg: google_protobuf_empty_pb.Empty = new google_protobuf_empty_pb.Empty()

        const getEnclavesPromise: Promise<Result<GetEnclavesResponse, Error>> = new Promise((resolve, _unusedReject) => {
            this.client.getEnclaves(emptyArg, (error: grpc.ServiceError | null, response?: GetEnclavesResponse) => {
                if (error === null) {
                    if (!response) {
                        resolve(err(new Error("No error was encountered but the response was still falsy; this should never happen")));
                    } else {
                        resolve(ok(response!));
                    }
                } else {
                    resolve(err(error));
                }
            })
        });

        const getEnclavesResult: Result<GetEnclavesResponse, Error> = await getEnclavesPromise;
        if (!getEnclavesResult.isOk()) {
            return err(getEnclavesResult.error)
        }

        const response: GetEnclavesResponse = getEnclavesResult.value;

        const enclavesMap: Map<string, EnclaveContext> = this.newEnclaveContextMapFromEnclaveInfoMap(response.getEnclaveInfoMap())

        return ok(enclavesMap)
    }

    public async stopEnclave(enclaveId: string): Promise<Result<null, Error>> {
        const args: StopEnclaveArgs = newStopEnclaveArgs(enclaveId)

        const stopEnclavePromise: Promise<Result<null, Error>> = new Promise((resolve, _unusedReject) => {
            this.client.stopEnclave(args, (error: Error | null, _unusedResponse?: google_protobuf_empty_pb.Empty) => {
                if (error === null) {
                    resolve(ok(null));
                } else {
                    resolve(err(error));
                }
            })
        });
        const stopEnclaveResult: Result<null, Error> = await stopEnclavePromise;
        if (!stopEnclaveResult.isOk()) {
            return err(stopEnclaveResult.error);
        }

        return ok(null);
    }

    public async destroyEnclave(enclaveId: string): Promise<Result<null, Error>> {
        const args: DestroyEnclaveArgs = newDestroyEnclaveArgs(enclaveId);

        const destroyEnclavePromise: Promise<Result<null, Error>> = new Promise((resolve, _unusedReject) => {
            this.client.destroyEnclave(args, (error: Error | null, _unusedResponse?: google_protobuf_empty_pb.Empty) => {
                if (error === null) {
                    resolve(ok(null));
                } else {
                    resolve(err(error));
                }
            })
        });
        const destroyEnclaveResult: Result<null, Error> = await destroyEnclavePromise;
        if (!destroyEnclaveResult.isOk()) {
            return err(destroyEnclaveResult.error);
        }

        return ok(null);
    }

    // ====================================================================================================
    //                                       Private helper functions
    // ====================================================================================================
    private newEnclaveContextMapFromEnclaveInfoMap(enclaveInfoMap: jspb.Map<string, EnclaveInfo>): Map<string, EnclaveContext> {

        const enclaveContextMap: Map<string, EnclaveContext> = new Map<string, EnclaveContext>()

        for (const [key, value] of enclaveInfoMap.entries()) {
            const newEnclaveContextResult: Result<EnclaveContext, Error> = this.newEnclaveContextFromEnclaveInfo(value);
            if (newEnclaveContextResult.isErr()) {
                // TODO This is really nasty - we skip enclaves that we can't create, and the only reason we wouldn't be able to create
                //  enclaves is because of stopped containers. Basically, we should move the API container into the engine-server, and
                //  make it impossible for enclaves to be in an invalid state at all
                continue;
            }
            enclaveContextMap.set(key, newEnclaveContextResult.value);
        }

        return enclaveContextMap;
    }

    private newEnclaveContextFromEnclaveInfo(enclaveInfo: EnclaveInfo): Result<EnclaveContext, Error> {
        const enclaveContainersStatus = enclaveInfo.getContainersStatus()
        if (enclaveContainersStatus !== EnclaveContainersStatus.ENCLAVECONTAINERSSTATUS_RUNNING) {
            return err(new Error(`Enclave containers status was '${enclaveContainersStatus}', but we can't create an enclave context from a non-running enclave`))
        }

        const enclaveApiContainerStatus = enclaveInfo.getApiContainerStatus()
        if (enclaveApiContainerStatus !== EnclaveAPIContainerStatus.ENCLAVEAPICONTAINERSTATUS_RUNNING) {
            return err(new Error(`Enclave API container status was '${enclaveApiContainerStatus}', but we can't create an enclave context without a running API container`))
        }

        const apiContainerInfo: EnclaveAPIContainerInfo | undefined = enclaveInfo.getApiContainerInfo();
        if (apiContainerInfo === undefined) {
            return err(new Error(`API container was listed as running, but no API container info exists`))
        }
        const apiContainerHostMachineInfo: EnclaveAPIContainerHostMachineInfo | undefined = enclaveInfo.getApiContainerHostMachineInfo()
        if (apiContainerHostMachineInfo === undefined) {
            return err(new Error(`API container was listed as running, but no API container host machine info exists`))
        }

        const apiContainerContext: ApiContainerContext = new ApiContainerContext(
            apiContainerInfo.getContainerId(),
            apiContainerInfo.getIpInsideEnclave(),
            apiContainerInfo.getPortInsideEnclave(),
            apiContainerHostMachineInfo.getIpOnHostMachine(),
            apiContainerHostMachineInfo.getPortOnHostMachine()
        )

        const enclaveContext: EnclaveContext = new EnclaveContext(
            enclaveInfo.getEnclaveId(),
            enclaveInfo.getNetworkId(),
            enclaveInfo.getNetworkCidr(),
            apiContainerContext,
        )

        return ok(enclaveContext);
    }
}
